sql/expression/mod.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/dolthub/vitess/go/mysql"
+	"github.com/shopspring/decimal"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+var _ ArithmeticOp = (*Mod)(nil)
+
+// Mod expression represents "%" arithmetic operation
+type Mod struct {
+	BinaryExpression
+}
+
+// NewMod creates a new Mod sql.Expression.
+func NewMod(left, right sql.Expression) *Mod {
+	return &Mod{BinaryExpression{Left: left, Right: right}}
+}
+
+func (m *Mod) LeftChild() sql.Expression {
+	return m.Left
+}
+
+func (m *Mod) RightChild() sql.Expression {
+	return m.Right
+}
+
+func (m *Mod) String() string {
+	return fmt.Sprintf("(%s %% %s)", m.Left, m.Right)
+}
+
+func (m *Mod) DebugString() string {
+	return fmt.Sprintf("(%s %% %s)", sql.DebugString(m.Left), sql.DebugString(m.Right))
+}
+
+// IsNullable implements the sql.Expression interface.
+func (m *Mod) IsNullable() bool {
+	if m.Type() == sql.Timestamp || m.Type() == sql.Datetime {
+		return true
+	}
+
+	return m.BinaryExpression.IsNullable()
+}
+
+// Type returns the greatest type for given operation.
+func (m *Mod) Type() sql.Type {
+	//TODO: what if both BindVars? should be constant folded
+	rTyp := m.Right.Type()
+	if sql.IsDeferredType(rTyp) {
+		return rTyp
+	}
+	lTyp := m.Left.Type()
+	if sql.IsDeferredType(lTyp) {
+		return lTyp
+	}
+
+	if isInterval(m.Left) || isInterval(m.Right) {
+		return sql.Datetime
+	}
+
+	if sql.IsTime(lTyp) && sql.IsTime(rTyp) {
+		return sql.Int64
+	}
+
+	// for division operation, it's either float or decimal.Decimal type
+	// except invalid value will result it either 0 or nil
+	return m.floatOrDecimal()
+}
+
+// floatOrDecimal returns either Float64 or decimaltype depending on column reference,
+// left and right expressions types and left and right evaluated types.
+// If there is float type column reference, the result type is always float
+// regardless of the column reference on the left or right side of division operation.
+// Otherwise, the return type is always decimal. The expression and evaluated types
+// are used to determine appropriate decimaltype to return that will not result in
+// precision loss.
+func (m *Mod) floatOrDecimal() sql.Type {
```

Comment:
I loved your suggestion to see if we can combine the two `floatOrDecimal` helper functions. I think it's worth trying out!